### DISCUSSION TITLE
What protocol should i use to discover peers across subnets in a local network?

### DISCUSSION THREAD

**Discussion by radiohertz** [libp2p/rust-libp2p#4222] [Q&A] [answered]
it seems mDNS doesn't really work across multiple subnets. Does this mean i have to go with [rendezvous](https://docs.libp2p.io/concepts/discovery-routing/rendezvous/) ?

**Comment by thomaseizinger** [comment]
From a networking PoV, it doesn't really make a difference whether you are in a single subnet, one hop away from the public internet or whether you have multiple local networks with different subnets (and routers).

Is your network running exclusively in these LANs or do you also expect nodes from outside this network to connect?

I assume you perform NAT between these networks? To enable direct connections, you will need to hole-punch between them (see dcutr). To _find_ them, you will need some kind of bootnode that is reachable by all nodes. Probably the easiest is if that bootnode runs kademlia and you make your own private DHT and perform random walks to find other nodes.

For a successful hole-punch, you will also need to have relay server(s).

How big do you envision your network to be?

**Reply by radiohertz** [reply]
> Is your network running exclusively in these LANs or do you also expect nodes from outside this network to connect?

we do expect nodes from outside the network to connect.

> I assume you perform NAT between these networks? To enable direct connections, you will need to hole-punch between them (see dcutr). To find them, you will need some kind of bootnode that is reachable by all nodes. Probably the easiest is if that bootnode runs kademlia and you make your own private DHT and perform random walks to find other nodes.

I see, thanks.

> How big do you envision your network to be?

Around 100 - 10,000.

**Reply by thomaseizinger** [reply, marked as answer]
> > Is your network running exclusively in these LANs or do you also expect nodes from outside this network to connect?
> 
> we do expect nodes from outside the network to connect.

Then there isn't really any difference from a topology PoV between your LANs and other networks.

> > How big do you envision your network to be?
> 
> Around 100 - 10,000.

So depending on where we are on that spectrum and how beefy your machines are, you could maintain full connectivity.

That makes things a bit easier. I'd suggest you deploy one or more, relatively strong machines to the public internet and use them as bootnodes. If they support kademlia, autonat and relay, all your other nodes can use them to connect amongst each other.

See https://github.com/mxinden/rust-libp2p-server.
